app/models/course/material.rb

```diff
@@ -42,6 +42,35 @@ class Course::Material < ApplicationRecord
 
   scope :in_concrete_folder, -> { joins(:folder).merge(Folder.concrete) }
 
+  class << self
+    def text_chunking!(material_ids, current_user)
+      materials = Course::Material.where(id: material_ids)
+      return if materials.empty?
+
+      materials.each(&:ensure_text_chunking!)
+      Course::Material::TextChunkJob.perform_later(material_ids, current_user).tap do |job|
+        materials.each do |material|
+          material.text_chunking.update_column(:job_id, job.job_id)
+        end
+      end
+    end
+
+    def destroy_text_chunk_references(material_ids)
+      ActiveRecord::Base.transaction do
+        materials = Course::Material.includes(:text_chunk_references).where(id: material_ids)
+        materials.each do |material|
+          # Only process materaial that are in 'chunked' state
+          return false if material.workflow_state != 'chunked'
```

Comment:
Why not filter from the DB query directly. filter materials where workflow_state is not chunked?